models/src/main/scala/coop/rchain/models/HashM.scala

```diff
@@ -0,0 +1,207 @@
+        connectiveUsedHash <- HashM[Boolean].hash(parSet.connectiveUsed)
+      } yield Objects.hash(Int.box(psHash), Int.box(remainderHash), Int.box(connectiveUsedHash))
+
+  }
+
+  implicit val parMapHash: HashM[ParMap] = new HashM[ParMap] {
+
+    override def hash[F[_]: Sync](parMap: ParMap): F[Int] =
+      for {
+        psHash             <- HashM[SortedParMap].hash(parMap.ps)
+        remainderHash      <- HashM[Option[Var]].hash(parMap.remainder)
+        connectiveUsedHash <- HashM[Boolean].hash(parMap.connectiveUsed)
+      } yield Objects.hash(Int.box(psHash), Int.box(remainderHash), Int.box(connectiveUsedHash))
+
+  }
+
+  implicit val BlockInfoHash                  = gen[BlockInfo]
+  implicit val BlockInfoWithoutTuplespaceHash = gen[BlockInfoWithoutTuplespace]
+  implicit val BlockQueryResponseHash         = gen[BlockQueryResponse]
+  implicit val ContinuationsWithBlockInfoHash = gen[ContinuationsWithBlockInfo]
+  implicit val DataWithBlockInfoHash          = gen[DataWithBlockInfo]
+  implicit val WaitingContinuationInfoHash    = gen[WaitingContinuationInfo]
+
+  implicit val ApprovedBlockHash             = gen[ApprovedBlock]
+  implicit val ApprovedBlockCandidateHash    = gen[ApprovedBlockCandidate]
+  implicit val BlockApprovalHash             = gen[BlockApproval]
+  implicit val BlockMessageHash              = gen[BlockMessage]
+  implicit val BodyHash                      = gen[Body]
+  implicit val BondHash                      = gen[Bond]
+  implicit val DeployHash                    = gen[Deploy]
+  implicit val DeployDataHash                = gen[DeployData]
+  implicit val FindDeployInBlockQueryHash    = gen[FindDeployInBlockQuery]
+  implicit val HeaderHash                    = gen[Header]
+  implicit val MaybeBlockMessageHash         = gen[MaybeBlockMessage]
+  implicit val ProcessedDeployHash           = gen[ProcessedDeploy]
+  implicit val RChainStateHash               = gen[RChainState]
+  implicit val UnapprovedBlockHash           = gen[UnapprovedBlock]
+  implicit val ListParWithRandomAndPhlosHash = gen[ListParWithRandomAndPhlos]
+
+  implicit val PCostHash              = gen[PCost]
+  implicit val TaggedContinuationHash = gen[TaggedContinuation]
+
+  implicit val PrivateNamePreviewQueryHash = gen[PrivateNamePreviewQuery]
+}
+
+trait HashMDerivation {
+  import magnolia._
+
+  type Typeclass[T] = HashM[T]
+
+  def combine[T](ctx: CaseClass[HashM, T]): HashM[T] = new HashM[T] {
+
+    def hash[F[_]: Sync](value: T): F[Int] = Sync[F].defer {
+      val hashFs = ctx.parameters.map { p =>
+        p.typeclass.hash(p.dereference(value))
+      }
+      for {
+        componentHashes <- hashFs.toList.sequence
+      } yield productHash(ctx.typeName.short, componentHashes)
+    }
+
+  }
+
+  //copied and adapted from scala.util.hashing.MurmurHash3,
+  //which is used in Scala's case class hash code implementation
+  private def productHash(prefix: String, elements: Seq[Int]): Int = {
```

Comment:
Probably overkill so feel free to ignore but maybe it's worth it to write a test that proofs that this is not generating too many collisions.

Reply:
The primary goal is to have it consistent with the default hashCode we had previously, even if it had excessive collisions (which I doubt). Also, since this is copied from scala std lib, I'd prefer to have at least a hint of experience that the collisions are excessive before testing / investigating / defining that :)